### PR TITLE
Refactor Accelerometer into an interface

### PR DESCRIPTION
**Pull request by antonfirsov**
## Description of the change

Contributes to #4497.

## Additions made

```C#
public interface IAccelerometer
{
	event EventHandler<AccelerometerChangedEventArgs>? ReadingChanged;
	event EventHandler? ShakeDetected;
	bool IsSupported { get; }
	bool IsMonitoring { get; }
	void Start(SensorSpeed sensorSpeed);
	void Stop();
}

public class AccelerometerImpl : IAccelerometer { /*... */ }
```

### CODE REVIEW

File: src/Essentials/src/Accelerometer/Accelerometer.tizen.cs [hunk 1]
```diff
@@ -3,28 +3,28 @@
...
+	public partial class AccelerometerImpl
 	{
 		internal static TizenAccelerometer DefaultSensor =>
 			(TizenAccelerometer)Platform.GetDefaultSensor(SensorType.Accelerometer);
 
 		internal static bool IsSupported =>
 			TizenAccelerometer.IsSupported;
 
-		static void PlatformStart(SensorSpeed sensorSpeed)
+		private void PlatformStart(SensorSpeed sensorSpeed)
```
Comment (rmarinho) [inline src/Essentials/src/Accelerometer/Accelerometer.tizen.cs:14]:
```suggestion
		void PlatformStart(SensorSpeed sensorSpeed)
```

File: src/Essentials/src/Accelerometer/Accelerometer.tizen.cs [hunk 2]
```diff
@@ -3,28 +3,28 @@
...
 		{
 			DefaultSensor.Interval = sensorSpeed.ToPlatform();
 			DefaultSensor.DataUpdated += DataUpdated;
 			DefaultSensor.Start();
 		}
 
-		static void PlatformStop()
+		private void PlatformStop()
```
Comment (rmarinho) [inline src/Essentials/src/Accelerometer/Accelerometer.tizen.cs:21]:
```suggestion
		void PlatformStop()
```

File: src/Essentials/src/Accelerometer/Accelerometer.tizen.cs [hunk 3]
```diff
@@ -3,28 +3,28 @@
...
 		{
 			DefaultSensor.DataUpdated -= DataUpdated;
 			DefaultSensor.Stop();
 		}
 
-		static void DataUpdated(object sender, AccelerometerDataUpdatedEventArgs e)
+		private void DataUpdated(object sender, AccelerometerDataUpdatedEventArgs e)
```
Comment (rmarinho) [inline src/Essentials/src/Accelerometer/Accelerometer.tizen.cs:27]:
```suggestion
		void DataUpdated(object sender, AccelerometerDataUpdatedEventArgs e)
```

File: src/Essentials/src/Accelerometer/Accelerometer.uwp.cs [hunk 4]
```diff
@@ -3,18 +3,18 @@
...
 
 		internal static WindowsAccelerometer DefaultSensor =>
 			WindowsAccelerometer.GetDefault();
 
-		internal static bool IsSupported =>
+		public bool IsSupported =>
 			DefaultSensor != null;
 
-		internal static void PlatformStart(SensorSpeed sensorSpeed)
+		private void PlatformStart(SensorSpeed sensorSpeed)
```
Comment (rmarinho) [inline src/Essentials/src/Accelerometer/Accelerometer.uwp.cs:17]:
```suggestion
		void PlatformStart(SensorSpeed sensorSpeed)
```

File: src/Essentials/src/Accelerometer/Accelerometer.uwp.cs [hunk 5]
```diff
@@ -24,14 +24,14 @@ internal static void PlatformStart(SensorSpeed sensorSpeed)
 			sensor.ReadingChanged += DataUpdated;
 		}
 
-		static void DataUpdated(object sender, AccelerometerReadingChangedEventArgs e)
+		private void DataUpdated(object sender, AccelerometerReadingChangedEventArgs e)
```
Comment (rmarinho) [inline src/Essentials/src/Accelerometer/Accelerometer.uwp.cs:27]:
```suggestion
		void DataUpdated(object sender, AccelerometerReadingChangedEventArgs e)
```

File: src/Essentials/src/Accelerometer/Accelerometer.uwp.cs [hunk 6]
```diff
@@ -24,14 +24,14 @@ internal static void PlatformStart(SensorSpeed sensorSpeed)
...
 		{
 			var reading = e.Reading;
 			var data = new AccelerometerData(reading.AccelerationX * -1, reading.AccelerationY * -1, reading.AccelerationZ * -1);
 			OnChanged(data);
 		}
 
-		internal static void PlatformStop()
+		private void PlatformStop()
```
Comment (rmarinho) [inline src/Essentials/src/Accelerometer/Accelerometer.uwp.cs:34]:
```suggestion
		void PlatformStop()
```

File: src/Essentials/src/Accelerometer/Accelerometer.shared.cs [hunk 7]
```diff
@@ -80,15 +76,15 @@ internal static void OnChanged(AccelerometerChangedEventArgs e)
 				ProcessShakeEvent(e.Reading.Acceleration);
 		}
 
-		static void ProcessShakeEvent(Vector3 acceleration)
+		private void ProcessShakeEvent(Vector3 acceleration)
```
Comment (rmarinho) [inline src/Essentials/src/Accelerometer/Accelerometer.shared.cs:79]:
```suggestion
		void ProcessShakeEvent(Vector3 acceleration)
```

File: src/Essentials/src/Accelerometer/Accelerometer.netstandard.tvos.macos.cs [hunk 8]
```diff
@@ -1,15 +1,15 @@
...
 	/// <include file="../../docs/Microsoft.Maui.Essentials/Accelerometer.xml" path="Type[@FullName='Microsoft.Maui.Essentials.Accelerometer']/Docs" />
-	public static partial class Accelerometer
+	public partial class AccelerometerImpl
 	{
-		internal static bool IsSupported =>
+		public bool IsSupported =>
 			throw ExceptionUtils.NotSupportedOrImplementedException;
 
-		static void PlatformStart(SensorSpeed sensorSpeed) =>
+		private void PlatformStart(SensorSpeed sensorSpeed) =>
```
Comment (rmarinho) [inline src/Essentials/src/Accelerometer/Accelerometer.netstandard.tvos.macos.cs:9]:
```suggestion
		void PlatformStart(SensorSpeed sensorSpeed) =>
```

File: src/Essentials/src/Accelerometer/Accelerometer.netstandard.tvos.macos.cs [hunk 9]
```diff
@@ -1,15 +1,15 @@
...
 			throw ExceptionUtils.NotSupportedOrImplementedException;
 
-		static void PlatformStop() =>
+		private void PlatformStop() =>
```
Comment (rmarinho) [inline src/Essentials/src/Accelerometer/Accelerometer.netstandard.tvos.macos.cs:12]:
```suggestion
		void PlatformStop() =>
```

File: src/Essentials/src/Accelerometer/Accelerometer.ios.watchos.cs [hunk 10]
```diff
@@ -25,7 +25,7 @@ static void DataUpdated(CMAccelerometerData data, NSError error)
 			OnChanged(accelData);
 		}
 
-		internal static void PlatformStop() =>
+		private void PlatformStop() =>
```
Comment (rmarinho) [inline src/Essentials/src/Accelerometer/Accelerometer.ios.watchos.cs:28]:
```suggestion
		void PlatformStop() =>
```

File: src/Essentials/src/Accelerometer/Accelerometer.ios.watchos.cs [hunk 11]
```diff
@@ -3,19 +3,19 @@
...
-		internal static void PlatformStart(SensorSpeed sensorSpeed)
+		private void PlatformStart(SensorSpeed sensorSpeed)
 		{
 			var manager = Platform.MotionManager;
 			manager.AccelerometerUpdateInterval = sensorSpeed.ToPlatform();
 			manager.StartAccelerometerUpdates(Platform.GetCurrentQueue(), DataUpdated);
 		}
 
-		static void DataUpdated(CMAccelerometerData data, NSError error)
+		private void DataUpdated(CMAccelerometerData data, NSError error)
```
Comment (rmarinho) [inline src/Essentials/src/Accelerometer/Accelerometer.ios.watchos.cs:18]:
```suggestion
		void DataUpdated(CMAccelerometerData data, NSError error)
```

File: src/Essentials/src/Accelerometer/Accelerometer.ios.watchos.cs [hunk 12]
```diff
@@ -3,19 +3,19 @@
...
 {
-	public static partial class Accelerometer
+	public partial class AccelerometerImpl
 	{
-		internal static bool IsSupported =>
+		public bool IsSupported =>
 			Platform.MotionManager?.AccelerometerAvailable ?? false;
 
-		internal static void PlatformStart(SensorSpeed sensorSpeed)
+		private void PlatformStart(SensorSpeed sensorSpeed)
```
Comment (rmarinho) [inline src/Essentials/src/Accelerometer/Accelerometer.ios.watchos.cs:11]:
```suggestion
		void PlatformStart(SensorSpeed sensorSpeed)
```

File: src/Essentials/src/Accelerometer/Accelerometer.android.cs [hunk 13]
```diff
@@ -3,23 +3,23 @@
...
 		{
 			var delay = sensorSpeed.ToPlatform();
-			listener = new AccelerometerListener();
+			listener = new AccelerometerListener(this);
 			accelerometer = Platform.SensorManager.GetDefaultSensor(SensorType.Accelerometer);
 			Platform.SensorManager.RegisterListener(listener, accelerometer, delay);
 		}
 
-		internal static void PlatformStop()
+		private void PlatformStop()
```
Comment (rmarinho) [inline src/Essentials/src/Accelerometer/Accelerometer.android.cs:22]:
```suggestion
		void PlatformStop()
```

File: src/Essentials/src/Accelerometer/Accelerometer.android.cs [hunk 14]
```diff
@@ -3,23 +3,23 @@
...
+		public bool IsSupported =>
 			Platform.SensorManager?.GetDefaultSensor(SensorType.Accelerometer) != null;
 
-		static AccelerometerListener listener;
-		static Sensor accelerometer;
+		AccelerometerListener listener;
+		Sensor accelerometer;
 
-		internal static void PlatformStart(SensorSpeed sensorSpeed)
+		private void PlatformStart(SensorSpeed sensorSpeed)
```
Comment (rmarinho) [inline src/Essentials/src/Accelerometer/Accelerometer.android.cs:14]:
```suggestion
		void PlatformStart(SensorSpeed sensorSpeed)
```

**Review by rmarinho — CHANGES_REQUESTED** [review]
Seems good, we just don't use `private` keyword , also we use use `_ ` on private fields, and we try not use `this.`

File: src/Essentials/src/Accelerometer/Accelerometer.android.cs [hunk 15]
```diff
@@ -35,8 +35,11 @@ class AccelerometerListener : Java.Lang.Object, ISensorEventListener
 		// acceleration due to gravity
 		const double gravity = 9.81;
 
-		internal AccelerometerListener()
+		private AccelerometerImpl accelerometer;
```
Comment (rmarinho) [inline src/Essentials/src/Accelerometer/Accelerometer.android.cs:38]:
```suggestion
		AccelerometerImpl accelerometer;
```

File: src/Essentials/src/Accelerometer/Accelerometer.android.cs [hunk 16]
```diff
@@ -35,8 +35,11 @@ class AccelerometerListener : Java.Lang.Object, ISensorEventListener
...
+
+		internal AccelerometerListener(AccelerometerImpl accelerometer)
 		{
+			this.accelerometer = accelerometer;
```
Comment (rmarinho) [inline src/Essentials/src/Accelerometer/Accelerometer.android.cs:42]:
```suggestion
			_accelerometer = accelerometer;
```

File: src/Essentials/src/Accelerometer/Accelerometer.android.cs [hunk 17]
```diff
@@ -49,7 +52,7 @@ void ISensorEventListener.OnSensorChanged(SensorEvent e)
 				return;
 
 			var data = new AccelerometerData(e.Values[0] / gravity, e.Values[1] / gravity, e.Values[2] / gravity);
-			Accelerometer.OnChanged(data);
+			accelerometer.OnChanged(data);
```
Comment (rmarinho) [inline src/Essentials/src/Accelerometer/Accelerometer.android.cs:55]:
```suggestion
			_accelerometer.OnChanged(data);
```

**Comment by antonfirsov** [comment]
@marinho thanks for pushing the style fixes! Is there something else I shall address?

Note that the Essentials project doesn't use the `_` prefix for fields in other types. Is it an inconsistency we should fix?

**Comment by rmarinho** [comment]
Yeah Essentials was merged to our project and i think we didn't fixe the style when we merged. I would say if you can fix as we go we should.

**Review by rmarinho — APPROVED** [review]

**Pull request merged**